rust/blockstore/src/arrow/block/types.rs

```diff
@@ -66,88 +66,119 @@ impl Block {
-        // and modified for two nested level comparisons.
+        key: Option<&K>,
+    ) -> usize {
         let mut size = self.len();
-        let mut left = 0;
-        let mut right = size;
-        let prefix_arr = self
+        if size == 0 {
+            return 0;
+        }
+
+        let prefix_array = self
             .data
             .column(0)
             .as_any()
             .downcast_ref::<StringArray>()
             .unwrap();
-        while left < right {
-            let mid = left + size / 2;
-
-            let prefix_cmp = prefix_arr.value(mid).cmp(prefix);
-            // This control flow produces conditional moves, which results in
-            // fewer branches and instructions than if/else or matching on
-            // cmp::Ordering.
-            // This is x86 asm for u8: https://rust.godbolt.org/z/698eYffTx.
-            left = if prefix_cmp == Less { mid + 1 } else { left };
-            right = if prefix_cmp == Greater { mid } else { right };
-            if prefix_cmp == Equal {
-                let key_cmp = K::get(self.data.column(1), mid)
-                    .partial_cmp(&key)
-                    .expect("NaN not expected"); // NaN not expected
-                left = if key_cmp == Less { mid + 1 } else { left };
-                right = if key_cmp == Greater { mid } else { right };
-                if key_cmp == Equal {
-                    return Some(V::get(self.data.column(2), mid));
-                }
+        let mut base = 0;
+
+        // This loop intentionally doesn't have an early exit if the comparison
+        // returns Equal. We want the number of loop iterations to depend *only*
+        // on the size of the input slice so that the CPU can reliably predict
+        // the loop count.
+        while size > 1 {
+            let half = size / 2;
+            let mid = base + half;
+
+            // SAFETY: the call is made safe by the following inconstants:
+            // - `mid >= 0`: by definition
+            // - `mid < size`: `mid = size / 2 + size / 4 + size / 8 ...`
+            let mut cmp = prefix_array.value(mid).cmp(prefix);
+
+            // Continue to compare the key if prefix matches
+            if let (Equal, Some(k)) = (cmp, key) {
+                cmp = K::get(self.data.column(1), mid)
+                    .partial_cmp(k)
+                    .expect("Array values should be comparable.");
             }
 
-            size = right - left;
+            base = if cmp == Less { mid } else { base };
+            size -= half;
+        }
+
+        // SAFETY: `base` is always in [0, size) because `base <= mid`.
+        // `base` should be the last index where the element is smaller than the target,
+        // or 0 if the first element is already larger than the target.
+        match prefix_array.value(base).cmp(prefix) {
+            Less => base + 1,
+            Equal => match key {
+                Some(k) => match K::get(self.data.column(1), base).partial_cmp(k) {
```

Comment:
nit: add a comment as why this is a partial_cmp and not a cmp

Reply:
Added comments